source/code/plugin/KubernetesApiClient.rb

```diff
@@ -1,128 +1,173 @@
-                           end    
-                       end
+    
+            def getClusterName
+                return @@ClusterName if !@@ClusterName.nil?
+                begin
+                    kubesystemResourceUri = "namespaces/" + @@KubeSystemNamespace + "/pods"
+                    podInfo = JSON.parse(getKubeResourceInfo(kubesystemResourceUri).body)
+                    podInfo['items'].each do |items|
+                        if items['metadata']['name'].include? "kube-controller-manager"
+                           items['spec']['containers'][0]['command'].each do |command|
+                               if command.include? "--cluster-name"
+                                   @@ClusterName = command.split('=')[1]
+                               end    
+                           end
+                        end
                     end
+                    @@ClusterName = "None"
+                rescue => error
+                    @Log.warn("cluster name request failed: #{error}")    
                 end
-                @@ClusterName = "None"
-            rescue => error
-                @Log.warn("cluster name request failed: #{error}")    
+                return @@ClusterName    
             end
-            return @@ClusterName    
-        end
-
-        def isNodeMaster
-            return @@IsNodeMaster if !@@IsNodeMaster.nil?
-            @@IsNodeMaster = false
-            begin
-                allNodesInfo =  getKubeResourceInfo('nodes')
-                if !allNodesInfo.nil? && !allNodesInfo.empty?
-                    thisNodeName = OMS::Common.get_hostname
-                    allNodesInfo['items'].each do |item|
-                        if item['metadata']['name'].casecmp(thisNodeName) == 0
-                            if !item['metadata']['labels'].to_s.include?("master") && item['metadata']['labels']["kubernetes.io/hostname"].to_s.split('-').last == '0'
-                                @@IsNodeMaster = true                                    
-                            end    
-                            break
-                        end
-                    end  
-                end         
-            rescue => error
-                @Log.warn("node role request failed: #{error}")    
+    
+            def isNodeMaster
+                return @@IsNodeMaster if !@@IsNodeMaster.nil?
+                @@IsNodeMaster = false
+                begin
+                    allNodesInfo = JSON.parse(getKubeResourceInfo('nodes').body)
+                    if !allNodesInfo.nil? && !allNodesInfo.empty?
+                        thisNodeName = OMS::Common.get_hostname
+                        allNodesInfo['items'].each do |item|
+                            if item['metadata']['name'].casecmp(thisNodeName) == 0
+                                if item['metadata']['labels']["kubernetes.io/role"].to_s.include?("master") && item['metadata']['labels']["kubernetes.io/hostname"].to_s.split('-').last == '0'
```

Comment:
Can you add a check for the Windows / Linux environment variable. Like we discussed the last time pinning it to the master doesnt work for pure linux environment, in which case it needs to be one of the agent nodes. For the time being we are pinning to node 0 .